File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import math

def solution(fees, records):
    answer = []
    basicTime,basicFee,unitTime,unitFee = map(int, fees)
    feeDic = {}
    timeDic = {}
    inRecords = {}
    
    def convertTime(time):
        hour,minutes = map(int, time.split(":"))
        return minutes + (hour * 60)
    
    def calcTime(inTime, outTime):
        return outTime - inTime
    
    def calcFee(time):
        if time <= basicTime:
            return basicFee
        else:
            return basicFee + (math.ceil((time - basicTime) / unitTime) * unitFee)
    
    for record in records: 
        time,carNumber,stat = record.split(" ")
        
        if stat == "IN":
            inRecords[carNumber] = time
        else:
            if(timeDic.get(carNumber) == None):
                timeDic[carNumber] = 0
            timeDic[carNumber] += calcTime(convertTime(inRecords[carNumber]), convertTime(time))
            inRecords[carNumber] = "0"
            
    # 미출차 시간 계산
    for num, time in inRecords.items():
        if time != "0":
            if(timeDic.get(num) == None):
                timeDic[num] = 0
            timeDic[num] += calcTime(convertTime(time), convertTime("23:59"))
    
    # 요금 계산
    for num, time in timeDic.items():
        if time != "0":
            if(feeDic.get(num) == None):
                feeDic[num] = 0
            feeDic[num] += calcFee(time)
    
    answer = [feeDic[key] for key in sorted(feeDic)]
        
    return answer

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.js
function solution(fees, records) {
    var answer = [];
    const [timeDefault,feeDefault,timeUnit,feeUnit] = fees
    const inInfo = {}
    const times = {}
    
    const timeToMin = (time) => {
        const [hh,mm] = time.split(":")
        return (+hh * 60) + (+mm)
    }
    
    const getCost = (totalTime) => {
        let time = totalTime - timeDefault
        if(time<0) time = 0
        
        return feeDefault + Math.ceil(time/timeUnit) * feeUnit
    }
    
    for(const record of records) {
        const [time,number,status] = record.split(" ")
        if(status === "IN") {
            inInfo[number] = timeToMin(time)
        } else {
            if(!times[number]) {
                times[number] = 0
            }
            
            times[number] += timeToMin(time) - inInfo[number]
            inInfo[number] = -1
        }
    }

    for(const [number,inTime] of Object.entries(inInfo)) {
        if(inTime > -1) {
            if(!times[number]) {
                times[number] = 0
            }
            
            times[number] += timeToMin('23:59') - inTime
        }
    }
    for(const [number,totalTime] of Object.entries(times)) {
       answer.push([number,getCost(totalTime)])
    }
    
    answer.sort((a,b) => a[0]-b[0])
    
    return answer.map(el => el[1]);
}